torch/csrc/jit/runtime/static/impl.cpp

```diff
@@ -333,6 +370,127 @@ StaticRuntime::IndividualMetrics StaticRuntime::benchmark_individual_ops(
+// supported.
+MemoryPlanner::MemoryPlanner(StaticRuntime* runtime) {
+  // collect register indices of outputs of ops with out variant
+  for (const ProcessedNode& node : runtime->nodes_) {
+    if (node.has_out_variant()) {
+      for (auto out : node.output_regs()) {
+        reg_out_variant_.insert(out);
+      }
+    }
+  }
+
+  // remove model outputs from reg_out_variant_
+  for (size_t output : runtime->module_->output_regs) {
+    reg_out_variant_.erase(output);
+  }
+
+  // remove tensors in output List/Tuple from reg_out_variant_
+  for (Value* output : runtime->module_->graph->outputs()) {
+    Node* output_node = output->node();
+    if (output_node->kind() == prim::TupleConstruct ||
+        output_node->kind() == prim::ListConstruct) {
+      for (Value* input : output_node->inputs()) {
+        reg_out_variant_.erase(runtime->module_->value_to_reg[input]);
+      }
+    }
+  }
+
+  // debug only
+  for (auto reg : reg_out_variant_) {
+    VLOG(1) << "reg_out_variant_: %"
+            << runtime->module_->values[reg]->debugName();
+  }
+}
+
+// Don't change the size if it is already aligned, otherwise increase the size
+// to make it aligned.
+size_t MemoryPlanner::compute_aligned_tensor_size(size_t nbytes) {
+  // Note: everything below is size_t
+  return (nbytes + c10::gAlignment - 1) & (~(c10::gAlignment - 1));
+}
+
+at::DataPtr MemoryPlanner::allocate_buffer(size_t size) {
+  at::Allocator* allocator = c10::GetCPUAllocator();
+  return allocator->allocate(size);
+}
+
+void MemoryPlanner::allocate_memory() {
+  if (internal_blob_max_sizes_sum_ == 0) {
+    return;
+  }
+
+  buffer_ = allocate_buffer(internal_blob_max_sizes_sum_);
+
+  size_t offset = 0;
+  uint8_t* start = static_cast<uint8_t*>(buffer_.get());
+
+  for (auto i = 0; i < internal_blob_storages_.size(); i++) {
+    auto tensor_size = internal_blob_max_sizes_[i];
+    if (tensor_size == 0) {
+      continue;
+    }
+    DCHECK_LE(offset + tensor_size, internal_blob_max_sizes_sum_);
+    void* src = static_cast<void*>(start + offset);
+
+    c10::StorageImpl* impl = internal_blob_storages_[i];
+    impl->set_data_ptr(at::DataPtr(src, src, nullptr, impl->device()));
+    impl->set_nbytes(tensor_size);
+
+    offset += tensor_size;
+  }
+  DCHECK_EQ(offset, internal_blob_max_sizes_sum_);
+}
+
+void MemoryPlanner::clean_up_memory(std::vector<IValue>& reg) {
+  if (internal_blob_storages_.empty() && !reg_out_variant_.empty()) {
+    // dedup tensor storages (tensor views share the same tensor storage)
+    std::unordered_set<c10::StorageImpl*> internal_blob_storages_set;
+    for (auto i : reg_out_variant_) {
+      internal_blob_storages_set.insert(
+          reg[i].toTensor().storage().unsafeGetStorageImpl());
+    }
+    internal_blob_storages_.assign(
+        internal_blob_storages_set.begin(), internal_blob_storages_set.end());
+
+    internal_blob_max_sizes_.resize(internal_blob_storages_.size());
+    reg_out_variant_.clear(); // no longer needed
+  }
```

Comment:
you might want to add some sanity check here that Storages don't change between iterations. General check is probably expensive, but you can run it in debug mode only or something (or on second iteration only). Basically if there's a bug somewhere in how we handle output argument, then reg_out_variant's element would get a new tensor storage assigned each time which won't be present in internal_blob_storages_.

Reply:
Will do